drivers/counter/counter_nrfx_rtc.c

```diff
@@ -81,18 +94,121 @@ static u32_t counter_nrfx_read(struct device *dev)
+		/* if top is not 2^n-1 */
+		u32_t diff;
+
+		diff = (val > old) ? (val - old) : val + top + 1 - old;
+
+		return diff;
+	}
+
+	return (val - old) & top;
+}
+
+/*
+ * Setting CC algorithm takes into account:
+ * - Procedure can be interrupted at any moment for multiple RTC ticks
+ * - Current COMPARE value written to the register may be close to the current
+ *   COUNTER value thus COMPARE event may be generated at any moment
+ * - Next COMPARE value may be soon in the future. Taking into accound potential
+ *   preemption COMPARE value may be set too late.
+ * - RTC registers are clocked with LF clock (32kHz)
+ * - Setting COMPARE register to COUNTER+1 may not generate COMPARE event
+ *
+ * Algorithm assumes that:
+ * - COMPARE interrupt is disabled
+ * - new value that is set is taking into account guard period. It means that
+ *   it won't be further in future than <top> - <guard_period> from now.
+ *
+ * Algorithm steps:
+ * - Read previous COMPARE value then write new value
+ * - Check if previous COMPARE value was located in near future
+ *   (between <COUNTER value> and <COUNTER value + guard period>. If yes, clear
+ *   COMPARE event. Note that new COMPARE value is already written so it may be
+ *   that COMPARE event for new value is cleared.
+ * - Check if new COMPARE value is one tick from current COUNTER value. If yes
+ *   busy wait for 1 LFCLK tick.
+ * - Check the relation between COUNTER value and COMPARE value. If COMPARE
+ *   value is in window between <COUNTER value - guard period> and current
+ *   COUNTER value and COMPARE event is not set then assume that COMPARE value
+ *   was set too late (e.g. due to preemption). Otherwise enable COMPARE
+ *   interrupt
+ */
+static int set_cc(struct device *dev, u8_t chan, u32_t val)
+{
+	const struct counter_nrfx_config *nrfx_config = get_nrfx_config(dev);
+	const nrfx_rtc_t *rtc = &nrfx_config->rtc;
+	NRF_RTC_Type  *reg = rtc->p_reg;
+	nrf_rtc_event_t evt;
+	u32_t prev_val;
+	u32_t top;
+	u32_t now;
+	u32_t diff;
+
+	/* todo Change to use nrfx API when added (it's coming). */
+	evt = (nrf_rtc_event_t)(offsetof(NRF_RTC_Type, EVENTS_COMPARE[chan]));
+
+	nrf_rtc_event_clear(reg, evt);
+	top = counter_get_top_value(dev);
+	now = nrf_rtc_counter_get(reg);
+	prev_val = nrf_rtc_cc_get(reg, chan);
+	nrf_rtc_cc_set(reg, chan, val);
+	diff = tick_diff(prev_val, now, top);
+
+	/* if previous CC was in short the future we need to take into account
+	 * that it may expire at any moment (given that current context can be
+	 * preempted at any instruction).
+	 */
+	if (diff < nrfx_config->guard_window) {
+		nrf_rtc_event_clear(reg, evt);
+	}
+
+	now = nrf_rtc_counter_get(reg);
+	diff = tick_diff(val, now, top);
+	if (diff == 1) {
+		/* Expire schedule for next tick, delay 1 tick, it's expected
+		 * that event will be set by then.
+		 * @todo what if prescaler is set
+		 */
+		LOG_INST_INF(nrfx_config->log,
+			"Delay introduced due to risk of pre-firing.");
+		NRFX_DELAY_US(LF_TICK_US);
```

Comment:
What about using the TICK event instead?

Reply:
Let me answer myself, it won't work that well with a prescaler.

Reply:
discussed offline with @anangl and fixed algorithm so that delay is no longer needed. Updated algorithm description.